service/src/main/java/com/commercetools/pspadapter/payone/notification/NotificationProcessorBase.java

```diff
@@ -53,29 +65,81 @@ public final void processTransactionStatusNotification(final Notification notifi
         }
 
         try {
-            client.executeBlocking(PaymentUpdateCommand.of(payment, createPaymentUpdates(payment, notification)));
+            // 1. update payment
+            CompletionStage<Order> fullCompletionStage = getPaymentService().updatePayment(payment, createPaymentUpdates(payment, notification))
+                    // 2. when payment is updated - try to find and updated respective Order.paymentState
+                    .thenComposeAsync(this::tryToUpdateOrderByPayment);
+
+            executeBlocking(fullCompletionStage);
+
         } catch (final io.sphere.sdk.client.ConcurrentModificationException e) {
             throw new java.util.ConcurrentModificationException(e);
         }
     }
 
     /**
+     * Fetch an order with respective payment id and try to update that order.
+     *
+     * @param updatedPayment <b>non-null</b> new updated payment instance.
+     * @return completion stage with nullable updated {@link Order}
+     */
+    private CompletionStage<Order> tryToUpdateOrderByPayment(Payment updatedPayment) {
+        return getOrderService().getOrderByPaymentId(updatedPayment.getId())
+                .thenComposeAsync(order -> updateOrderIfExists(order.orElse(null), updatedPayment));
+    }
+
+    /**
+     * If the {@code order} is not null and it's payment status is different from the new one - update it,
+     * otherwise - skip updating.
+     * <p>
+     * If a new payment status can't be mapped to the order's payment state - the value is left unchanged.
+     *
+     * @param order          <b>nullable</b> order to update state. If <b>null</b> - the update is skipped,
+     *                       but the incident is reported.
+     * @param updatedPayment <b>non-null</b> instance of the updated payment
+     * @return a {@link CompletionStage} with <b>nullable</b> updated {@link Order} instance. If incoming order is null
+     * or new payment state is undefined, or update is not required - the returned stage is completed and contains
+     * the same instance.
+     */
+    private CompletionStage<Order> updateOrderIfExists(Order order, Payment updatedPayment) {
+        if (order != null) {
```

Comment:
Kind of holly-war topic, but at least the main developer of Java `Optional` recommends not to use it as a method argument:

> You should almost never use it as a field of something or a method parameter

http://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555